File: ServerScripts/python-cli/add-service.py
import ipaddress
import os
import json
import random

class LanguageSetup():
    def __init__(self):
        self.languageNumberMapping = {
            "0" : "Exit",
            "1" : "Python"
        }
        self.languageNumber = None
        self.languageName = None

    def selectLanguage(self):
        print("Select the Language for the Service")
        print()
        for lang in self.languageNumberMapping.keys():
            print(f"{lang} : {self.languageNumberMapping[lang]}") 

        while True:
            print()
            language = input("Enter the Language Number: ")

            if language == "0":
                exit(0)
            
            if not language.isnumeric():
                print("Invalid Input. Please enter a valid number.")
                continue
        
            if language not in self.languageNumberMapping.keys():
                print("Invalid Language Number. Please try again.")
                continue

            print("\n\n--------------------------------------------------------------\n\n")

            self.languageNumber = int(language)
            self.languageName = self.languageNumberMapping[language]
            
            return language

class PythonTemplateSetup():
    def __init__(self):
        self.serviceName = None
        self.serviceHttpHost = None
        self.serviceHttpPort = None
        self.servicePrivilegedIpAddresses = []
        self.enableCors = True  # Default: CORS enabled

    def getServiceName(self):
        print("Enter the Name of the Service")
        while True:
            print()
            serviceName = input("Enter the Service Name: ")
            serviceName = serviceName.strip()

            if " " in serviceName:
                print("Service Name should not contain spaces. Please enter a valid name without spaces.")
                continue
        
            if "-" in serviceName:
                print("Service Name should not contain dashes. Please enter a valid name without dashes.")
                continue

            if "/" in serviceName:
                print("Service Name should not contain slashes. Please enter a valid name without slashes.")
                continue

            if "," in serviceName:
                print("Service Name should not contain commas. Please enter a valid name without commas.")
                continue

            if serviceName == "":
                print("Service Name cannot be empty. Please enter a valid name.")
                continue

            print("\n\n--------------------------------------------------------------\n\n")
            return serviceName

    def getHostandPortForHttpServer(self):
        print("Enter the Host and Port for the HTTP Server")

        httpServerHost = None
        httpServerPort = None

        while True:
            print("Enter the Host for the HTTP Server (Default: 127.0.0.1)")

            httpServerHost = input("Host: ")
            if httpServerHost == "":
                httpServerHost = "127.0.0.1"

            try:
                # Try to validate as an IP address
                ipaddress.ip_address(httpServerHost)
            except ValueError:
                if httpServerHost != "localhost":
                    print("Invalid host. Please enter a valid IP address or 'localhost'.")
                    continue
            
            print()
            break
    
        print("Enter the Port for the HTTP Server (Default: random)")
        while True:
            httpServerPort = input("Port: ")
            if httpServerPort == "":
                httpServerPort = random.randint(1024, 65535)
                break

            if not httpServerPort.isnumeric():
                print("Invalid port. Please enter a valid number.")
                print()
                continue
            
            httpServerPort = int(httpServerPort)
            if httpServerPort < 1024 or httpServerPort > 65535:
                print("Port number must be between 1024 and 65535. Please try again.")
                print()
                continue
            
            break

        print("\n\n--------------------------------------------------------------\n\n")
        return httpServerHost, httpServerPort

    def getPrivilegedIpAddresses(self):
        print("Enter Privileged IP Addresses (one per line, press Enter on empty line to finish)")
        print("Default: 127.0.0.1 will be added automatically")
        
        privilegedIps = ["127.0.0.1"]  # Default privileged IP
        
        while True:
            ipInput = input("Enter IP Address (or press Enter to finish): ").strip()
            
            if ipInput == "":
                break
                
            try:
                # Validate IP address
                ipaddress.ip_address(ipInput)
                if ipInput not in privilegedIps:
                    privilegedIps.append(ipInput)
                    print(f"Added: {ipInput}")
                else:
                    print(f"IP {ipInput} already in list")
            except ValueError:
                print("Invalid IP address. Please enter a valid IP address.")
                continue
        
        print("\n\n--------------------------------------------------------------\n\n")
        return privilegedIps

    def printServiceConfiguration(self):
        print("=== SERVICE CONFIGURATION ===")
        print(f"Service Name: {self.serviceName}")
        print(f"HTTP Host: {self.serviceHttpHost}")
        print(f"HTTP Port: {self.serviceHttpPort}")
        print(f"Privileged IP Addresses: {self.servicePrivilegedIpAddresses}")
        print("=============================")

    def createServiceDirectory(self):
        # Create service folder name
        serviceFolderName = f"service_{self.serviceName}Service"
        
        # Get grandparent directory (two levels up from current script)
        currentDir = os.path.dirname(os.path.abspath(__file__))
        parentDir = os.path.dirname(currentDir)
        grandparentDir = os.path.dirname(parentDir)
        
        # Create full path for service directory
        serviceDirPath = os.path.join(grandparentDir, serviceFolderName)
        
        # Create directory if it doesn't exist
        os.makedirs(serviceDirPath, exist_ok=True)
        
        return serviceDirPath

    def generateServiceFile(self, serviceDirPath):
        # Create service file name
        serviceFileName = f"{self.serviceName.lower()}-service.py"
        serviceFilePath = os.path.join(serviceDirPath, serviceFileName)
        
        # Read the HTTP service template from the correct location
        currentDir = os.path.dirname(os.path.abspath(__file__))
        parentDir = os.path.dirname(currentDir)
        templatePath = os.path.join(parentDir, "ServiceTemplates", "python", "HTTP_SERVICE.txt")
        
        with open(templatePath, 'r') as templateFile:
            templateContent = templateFile.read()
        
        # Replace placeholders with actual values
        modifiedContent = self.replacePlaceholders(templateContent)
        
        # Write the modified content to the new service file
        with open(serviceFilePath, 'w') as serviceFile:
            serviceFile.write(modifiedContent)
        
        return serviceFilePath

    def replacePlaceholders(self, templateContent):
        # Replace the host
        templateContent = self.replaceSection(
            templateContent,
            "#<HTTP_SERVER_HOST_START>",
            "#<HTTP_SERVER_HOST_END>",
            f'    httpServerHost = "{self.serviceHttpHost}"'
        )
        
        # Replace the port
        templateContent = self.replaceSection(
            templateContent,
            "#<HTTP_SERVER_PORT_START>",
            "#<HTTP_SERVER_PORT_END>",
            f'    httpServerPort = {self.serviceHttpPort}'
        )
        
        # Replace privileged IP addresses
        privilegedIpsStr = "[" + ", ".join([f'"{ip}"' for ip in self.servicePrivilegedIpAddresses]) + "]"
        templateContent = self.replaceSection(
            templateContent,
            "#<HTTP_SERVER_PRIVILEGED_IP_ADDRESS_START>",
            "#<HTTP_SERVER_PRIVILEGED_IP_ADDRESS_END>",
            f'    httpServerPrivilegedIpAddress = {privilegedIpsStr}'
        )

        # Replace CORS middleware section
        cors_line = 'self.app.add_middleware(CORSMiddleware, allow_origins=["*"],allow_credentials=True,allow_methods=["*"],allow_headers=["*"],)'
        if self.enableCors:
            cors_code = f'        {cors_line}'
        else:
            cors_code = f'        #{cors_line}'
        templateContent = self.replaceSection(
            templateContent,
            "#<HTTP_SERVER_CORS_ADDITION_START>",
            "#<HTTP_SERVER_CORS_ADDITION_END>",
            cors_code
        )
        
        return templateContent

    def replaceSection(self, content, startMarker, endMarker, newContent):
        startIndex = content.find(startMarker)
        if startIndex == -1:
            return content
        
        endIndex = content.find(endMarker)
        if endIndex == -1:
            return content
        
        # Find the start of the start marker line to get proper indentation
        lineStart = content.rfind('\n', 0, startIndex) + 1
        indentation = content[lineStart:startIndex]
        
        # Find the end of the end marker line
        endLineEnd = content.find('\n', endIndex)
        if endLineEnd == -1:
            endLineEnd = len(content)
        else:
            endLineEnd += 1  # Include the newline
        
        # Replace the section with proper indentation for end marker
        before = content[:startIndex]
        after = content[endLineEnd:]
        
        return before + startMarker + '\n' + newContent + '\n' + indentation + endMarker + '\n' + after

    def updateServicesJson(self):
        # Get the path to services.json
        currentDir = os.path.dirname(os.path.abspath(__file__))
        parentDir = os.path.dirname(os.path.dirname(currentDir))  # Go up two levels to reach /home/paarth/Test
        servicesJsonPath = os.path.join(parentDir, "services.json")
        
        # Read existing services
        services = []
        if os.path.exists(servicesJsonPath):
            with open(servicesJsonPath, 'r') as file:
                services = json.load(file)
        
        # Create new service entry
        newService = {
            "ServiceLanguage": "Python",
            "ServiceName": self.serviceName,
            "ServiceFolderName": f"service_{self.serviceName}Service",
            "ServiceFileName": f"{self.serviceName.lower()}-service.py",
            "ServiceHttpHost": self.serviceHttpHost,
            "ServiceHttpPriviledgedIpAddress": self.servicePrivilegedIpAddresses,
            "ServiceHttpPort": self.serviceHttpPort,
            "ServiceType": "HTTP_SERVICE"
        }
        
        # Add new service to the list
        services.append(newService)
        
        # Write back to file
        with open(servicesJsonPath, 'w') as file:
            json.dump(services, file, indent=4)
        
        print(f"Updated services.json with new service: {self.serviceName}")

    def updateEnvFile(self):
        # Get the path to .env file
        currentDir = os.path.dirname(os.path.abspath(__file__))
        parentDir = os.path.dirname(os.path.dirname(currentDir))  # Go up two levels to reach /home/paarth/Test
        envFilePath = os.path.join(parentDir, ".env")
        
        # Read existing .env content
        envContent = ""
        if os.path.exists(envFilePath):
            with open(envFilePath, 'r') as file:
                envContent = file.read()
        
        # Prepare service entry
        serviceEntry = f'{self.serviceName.upper()}_SERVICE = "{self.serviceHttpHost}:{self.serviceHttpPort}"'
        commentedServiceEntry = f'# {self.serviceName.upper()}_SERVICE = "{self.serviceHttpHost}:{self.serviceHttpPort}"'
        
        # Add to development section
        devMarker = "#<ADD_DEVELOPMENT_SERVICES_ENVRIONMENT_VARIABLES>"
        if devMarker in envContent:
            envContent = envContent.replace(devMarker, f"{devMarker}\n{serviceEntry}")
        
        # Add to production section (commented)
        prodMarker = "#<ADD_PRODUCTION_SERVICES_ENVRIONMENT_VARIABLES>"
        if prodMarker in envContent:
            envContent = envContent.replace(prodMarker, f"{prodMarker}\n{commentedServiceEntry}")
        
        # Write back to file
        with open(envFilePath, 'w') as file:
            file.write(envContent)
        
        print(f"Updated .env file with new service environment variables")

    def askEnableCors(self):
        print("Enable CORS middleware for this service? (Y/n) [Default: Y]")
        while True:
            answer = input("Enable CORS? (Y/n): ").strip().lower()
            if answer == "" or answer == "y":
                return True
            if answer == "n":
                return False
            print("Invalid input. Please enter 'Y' or 'n'.")

    def startServiceSetup(self):
        self.serviceName = self.getServiceName()
        self.serviceHttpHost, self.serviceHttpPort = self.getHostandPortForHttpServer()
        self.servicePrivilegedIpAddresses = self.getPrivilegedIpAddresses()
        self.enableCors = self.askEnableCors()
        
        self.printServiceConfiguration()
        
        # Create service directory and file
        serviceDirPath = self.createServiceDirectory()
        serviceFilePath = self.generateServiceFile(serviceDirPath)
        
        # Update services.json and .env files
        self.updateServicesJson()
        self.updateEnvFile()
        
        print(f"\nService created successfully!")
        print(f"Service Directory: {serviceDirPath}")
        print(f"Service File: {serviceFilePath}")
        print(f"Updated configuration files: services.json and .env")

langSetup = LanguageSetup()
langSetup.selectLanguage()

if langSetup.languageNumber == 1:
    pythonTemplateSetup = PythonTemplateSetup()
    pythonTemplateSetup.startServiceSetup()

File: ServerScripts/js-cli/add-service.js
const fs = require('fs');
const path = require('path');
const readline = require('readline');
const os = require('os');

function askQuestion(query) {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });
    return new Promise(resolve => rl.question(query, ans => {
        rl.close();
        resolve(ans);
    }));
}

async function getServiceName() {
    console.log("Enter the Name of the Service");
    while (true) {
        console.log();
        let serviceName = (await askQuestion("Enter the Service Name: ")).trim();
        if (!serviceName) {
            console.log("Service Name cannot be empty. Please enter a valid name.");
            continue;
        }
        if (serviceName.includes(" ")) {
            console.log("Service Name should not contain spaces. Please enter a valid name without spaces.");
            continue;
        }
        if (serviceName.includes("-")) {
            console.log("Service Name should not contain dashes. Please enter a valid name without dashes.");
            continue;
        }
        if (serviceName.includes("/")) {
            console.log("Service Name should not contain slashes. Please enter a valid name without slashes.");
            continue;
        }
        if (serviceName.includes(",")) {
            console.log("Service Name should not contain commas. Please enter a valid name without commas.");
            continue;
        }
        console.log("\n\n--------------------------------------------------------------\n\n");
        return serviceName;
    }
}

function isValidIpAddress(ip) {
    const ipv4Regex = /^(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)$/;
    const ipv6Regex = /^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))$/;
    
    return ipv4Regex.test(ip) || ipv6Regex.test(ip);
}

async function getHostAndPort() {
    console.log("Enter the Host and Port for the HTTP Server");
    
    let host;
    while (true) {
        console.log("Enter the Host for the HTTP Server (Default: 127.0.0.1)");
        host = await askQuestion("Host: ");
        if (!host) {
            host = "127.0.0.1";
            break;
        }
        if (isValidIpAddress(host) || host === "localhost") {
            break;
        }
        console.log("Invalid host. Please enter a valid IP address or 'localhost'.");
        console.log();
    }
    
    console.log();
    console.log("Enter the Port for the HTTP Server (Default: random)");
    let port;
    while (true) {
        let portStr = await askQuestion("Port: ");
        if (!portStr) {
            port = Math.floor(Math.random() * (65535 - 1024 + 1)) + 1024;
            break;
        }
        
        if (!/^\d+$/.test(portStr)) {
            console.log("Invalid port. Please enter a valid number.");
            console.log();
            continue;
        }
        
        port = parseInt(portStr);
        if (port < 1024 || port > 65535) {
            console.log("Port number must be between 1024 and 65535. Please try again.");
            console.log();
            continue;
        }
        
        break;
    }
    
    console.log("\n\n--------------------------------------------------------------\n\n");
    return { host, port };
}

async function getPrivilegedIps() {
    console.log("Enter Privileged IP Addresses (one per line, press Enter on empty line to finish)");
    console.log("Default: 127.0.0.1 will be added automatically");
    let ips = ["127.0.0.1"];
    while (true) {
        let ip = (await askQuestion("Enter IP Address (or press Enter to finish): ")).trim();
        if (!ip) break;
        
        if (!isValidIpAddress(ip)) {
            console.log("Invalid IP address. Please enter a valid IPv4 or IPv6 address.");
            continue;
        }
        
        if (!ips.includes(ip)) {
            ips.push(ip);
            console.log(`Added: ${ip}`);
        } else {
            console.log(`IP ${ip} already in list`);
        }
    }
    console.log("\n\n--------------------------------------------------------------\n\n");
    return ips;
}

async function askEnableCors() {
    console.log("Enable CORS middleware for this service? (Y/n) [Default: Y]");
    while (true) {
        let ans = (await askQuestion("Enable CORS? (Y/n): ")).trim().toLowerCase();
        if (ans === "" || ans === "y") {
            return true;
        }
        if (ans === "n") {
            return false;
        }
        console.log("Invalid input. Please enter 'Y' or 'n'.");
    }
}

function createServiceDirectory(serviceName) {
    const currentDir = __dirname;
    const grandparentDir = path.dirname(path.dirname(currentDir));
    const serviceFolderName = `service_${serviceName}Service`;
    const serviceDirPath = path.join(grandparentDir, serviceFolderName);
    if (!fs.existsSync(serviceDirPath)) fs.mkdirSync(serviceDirPath, { recursive: true });
    return serviceDirPath;
}

function generateServiceFile(serviceDirPath, serviceName, host, port, privilegedIps, enableCors) {
    const currentDir = __dirname;
    const templatePath = path.join(path.dirname(currentDir), "ServiceTemplates", "Js", "HTTP_SERVICE.txt");
    
    if (!fs.existsSync(templatePath)) {
        console.error(`Template file not found: ${templatePath}`);
        console.error("Please ensure the template exists at the correct location.");
        return null;
    }
    
    let templateContent = fs.readFileSync(templatePath, 'utf-8');

    // Replace placeholders in the template - more specific replacements
    templateContent = templateContent.replace(
        'const httpServerHost = "127.0.0.1";',
        `const httpServerHost = "${host}";`
    );
    
    templateContent = templateContent.replace(
        'const httpServerPort = 8080;',
        `const httpServerPort = ${port};`
    );
    
    templateContent = templateContent.replace(
        'const httpServerPrivilegedIpAddress = ["127.0.0.1"];',
        `const httpServerPrivilegedIpAddress = [${privilegedIps.map(ip => `"${ip}"`).join(", ")}];`
    );
    
    // Handle CORS configuration - replace the entire CORS middleware section
    if (!enableCors) {
        templateContent = templateContent.replace(
            /this\.app\.use\(cors\(\{[\s\S]*?\}\)\);/,
            '// CORS disabled by user choice\n        // this.app.use(cors({ origin: \'*\', credentials: true, methods: \'*\', allowedHeaders: \'*\' }));'
        );
    }

    const serviceFileName = `${serviceName.toLowerCase()}-service.js`;
    const serviceFilePath = path.join(serviceDirPath, serviceFileName);
    
    try {
        fs.writeFileSync(serviceFilePath, templateContent);
        console.log(`Service file created: ${serviceFilePath}`);
        return serviceFilePath;
    } catch (error) {
        console.error(`Error creating service file: ${error.message}`);
        return null;
    }
}

function updateServicesJson(serviceName, host, port, privilegedIps) {
    const currentDir = __dirname;
    const servicesJsonPath = path.join(path.dirname(path.dirname(currentDir)), "services.json");
    let services = [];
    if (fs.existsSync(servicesJsonPath)) {
        try {
            services = JSON.parse(fs.readFileSync(servicesJsonPath, 'utf-8'));
        } catch (error) {
            console.log("Error reading services.json, creating new array");
            services = [];
        }
    }
    
    const newService = {
        "ServiceLanguage": "JavaScript",
        "ServiceName": serviceName,
        "ServiceFolderName": `service_${serviceName}Service`,
        "ServiceFileName": `${serviceName.toLowerCase()}-service.js`,
        "ServiceHttpHost": host,
        "ServiceHttpPrivilegedIpAddress": privilegedIps,
        "ServiceHttpPort": port,
        "ServiceWsPort": null,
        "ServiceType": "HTTP_SERVICE"
    };
    
    services.push(newService);
    fs.writeFileSync(servicesJsonPath, JSON.stringify(services, null, 4));
    console.log(`Updated services.json with new service: ${serviceName}`);
}

function updateEnvFile(serviceName, host, port) {
    const currentDir = __dirname;
    const envFilePath = path.join(path.dirname(path.dirname(currentDir)), ".env");
    let envContent = "";
    if (fs.existsSync(envFilePath)) {
        envContent = fs.readFileSync(envFilePath, 'utf-8');
    }
    const serviceEntry = `${serviceName.toUpperCase()}_SERVICE = "${host}:${port}"`;
    const commentedServiceEntry = `# ${serviceName.toUpperCase()}_SERVICE = "${host}:${port}"`;
    const devMarker = "#<ADD_DEVELOPMENT_SERVICES_ENVRIONMENT_VARIABLES>";
    const prodMarker = "#<ADD_PRODUCTION_SERVICES_ENVRIONMENT_VARIABLES>";
    if (envContent.includes(devMarker)) {
        envContent = envContent.replace(devMarker, `${devMarker}\n${serviceEntry}`);
    }
    if (envContent.includes(prodMarker)) {
        envContent = envContent.replace(prodMarker, `${prodMarker}\n${commentedServiceEntry}`);
    }
    fs.writeFileSync(envFilePath, envContent);
    console.log(`Updated .env file with new service environment variables`);
}

async function main() {
    console.log("=== JavaScript HTTP Service Setup ===");
    const serviceName = await getServiceName();
    const { host, port } = await getHostAndPort();
    const privilegedIps = await getPrivilegedIps();
    const enableCors = await askEnableCors();

    console.log("=== SERVICE CONFIGURATION ===");
    console.log(`Service Name: ${serviceName}`);
    console.log(`HTTP Host: ${host}`);
    console.log(`HTTP Port: ${port}`);
    console.log(`Privileged IP Addresses: ${privilegedIps.join(", ")}`);
    console.log("=============================");

    const serviceDirPath = createServiceDirectory(serviceName);
    const serviceFilePath = generateServiceFile(serviceDirPath, serviceName, host, port, privilegedIps, enableCors);

    updateServicesJson(serviceName, host, port, privilegedIps);
    updateEnvFile(serviceName, host, port);

    console.log(`\nService created successfully!`);
    console.log(`Service Directory: ${serviceDirPath}`);
    console.log(`Service File: ${serviceFilePath}`);
    console.log(`Updated configuration files: services.json and .env`);
}

if (require.main === module) {
    main();
}
